### PR TITLE
Method for 'o' packet (Checking if a server is OMP)

### PR DESCRIPTION
**Pull request by markski1**
### Description of change

- Adds a method for checking if a server is using open.mp or not, through a special 'o' packet.
- Does some corrections for compliance with modern .NET (stuff not marked as nullable cannot be initialized as null, etc.)
- Other changes to fix compilation issues in modern .NET (also mostly related to handling nullability in some parsing bits.)
- Small formatting changes.

Controvertial change:
- 'TryParseByte' helper, which returns a "0" on unparseable values.

This is controvertial, because returning a 0 for values that can't be parsed is illogical. The problem is that the current behaviour in these cases is crashing, and I'm not sure how else to handle it. I had certain "random" crashes on SAMonitor which stopped when I implemented this.

BE ADVISED: For some reason with my VS setup, I could not run the tests, but this version of SampQuery is currently running on [SAMonitor](https://sam.markski.ar). So I'm quite confident everything works correctly, but if someone can do the tests on their machine for me I guess that'd be better.

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [ ] All tests are passed (if tests fail due to timeout, try running the tests later or run them sequentially)
- [ ] This pull request links relevant issues as `Fixes #0000`
- [ ] There are new or updated unit tests validating the change
- [x] Documentation (README.md) has been updated to reflect this change

### CODE REVIEW

**Comment by markski1** [comment]
Just remembered I didn't make the changes to README, and the OMP method wrongly claims to be asynchronous. I'll fix those when I'm on my laptop.

File: Src/SampQuery.cs [hunk 1]
```diff
@@ -382,15 +444,15 @@ private ServerRules CollectServerRulesFromByteArray(byte[] data) {
...
+                        PropertyInfo? property = sampServerRulesData.GetType().GetProperty(new string(read.ReadChars(read.ReadByte())).Replace(' ', '_'), BindingFlags.IgnoreCase | BindingFlags.Public | BindingFlags.Instance);
                         value = new string(read.ReadChars(read.ReadByte()));
 
                         if (property != null)
                         {
                             if (property.PropertyType == typeof(bool)) val = value == "On";
                             else if (property.PropertyType == typeof(Uri)) val = Helpers.ParseWebUrl(value);
                             else if (property.PropertyType == typeof(DateTime)) val = Helpers.ParseTime(value);
-                            else val = Convert.ChangeType(value, property.PropertyType, CultureInfo.InvariantCulture);
+                            else val = Helpers.TryParseByte(value, property);
```
Comment (grish-st) [inline Src/SampQuery.cs:455]: This block should be refactored, properties need to be checked by name and not by type. What about `TryParseByte`, it should be tested and unit tests should be written.

File: Src/Types/ServerInfo.cs [hunk 2]
```diff
@@ -8,36 +8,36 @@ public class ServerInfo
         /// <summary>
         /// Hostname
         /// </summary>
-        public string HostName { get; set; } 
+        public string HostName { get; set; } = "UNKNOWN";
```
Comment (grish-st) [inline Src/Types/ServerInfo.cs:11]: Why?

File: Src/Types/ServerInfo.cs [hunk 3]
```diff
@@ -8,36 +8,36 @@ public class ServerInfo
...
 
         /// <summary>
         /// Gamemode text
         /// </summary>
-        public string GameMode { get; set; }
+        public string GameMode { get; set; } = "UNKNOWN";
```
Comment (grish-st) [inline Src/Types/ServerInfo.cs:16]: Why?

File: Src/Types/ServerInfo.cs [hunk 4]
```diff
@@ -8,36 +8,36 @@ public class ServerInfo
...
         /// Server language 
         /// </summary>
-        public string Language { get; set; }
+        public string Language { get; set; } = "UNKNOWN";
 
         /// <summary>
         /// Number of players online
         /// </summary>
-        public ushort Players { get; set; }
+        public ushort Players { get; set; } = 0;
```
Comment (grish-st) [inline Src/Types/ServerInfo.cs:26]: Why?

File: Src/Types/ServerInfo.cs [hunk 5]
```diff
@@ -8,36 +8,36 @@ public class ServerInfo
...
         /// Gamemode text
         /// </summary>
-        public string GameMode { get; set; }
+        public string GameMode { get; set; } = "UNKNOWN";
 
         /// <summary>
         /// Server language 
         /// </summary>
-        public string Language { get; set; }
+        public string Language { get; set; } = "UNKNOWN";
```
Comment (grish-st) [inline Src/Types/ServerInfo.cs:21]: Why?

File: Src/Types/ServerInfo.cs [hunk 6]
```diff
@@ -8,36 +8,36 @@ public class ServerInfo
...
         /// Number of players online
         /// </summary>
-        public ushort Players { get; set; }
+        public ushort Players { get; set; } = 0;
 
         /// <summary>
         /// Maximum number of players 
         /// </summary>
-        public ushort MaxPlayers { get; set; }
+        public ushort MaxPlayers { get; set; } = 0;
```
Comment (grish-st) [inline Src/Types/ServerInfo.cs:31]: Why?

File: Src/Types/ServerInfo.cs [hunk 7]
```diff
@@ -8,36 +8,36 @@ public class ServerInfo
...
 
         /// <summary>
         /// Password availability
         /// </summary>
-        public bool Password { get; set; }
+        public bool Password { get; set; } = false;
```
Comment (grish-st) [inline Src/Types/ServerInfo.cs:36]: Why?

File: Src/Types/ServerInfo.cs [hunk 8]
```diff
@@ -8,36 +8,36 @@ public class ServerInfo
...
 
         /// <summary>
         /// Ping of server
         /// </summary>
-        public int ServerPing { get; set; }
+        public int ServerPing { get; set; } = -1;
```
Comment (grish-st) [inline Src/Types/ServerInfo.cs:41]: Why?

File: Src/Types/ServerPlayer.cs [hunk 9]
```diff
@@ -13,7 +13,7 @@ public class ServerPlayer
         /// <summary>
         /// Player Name. 
         /// </summary>
-        public string PlayerName { get; set; }
+        public string PlayerName { get; set; } = "UNKNOWN";
```
Comment (grish-st) [inline Src/Types/ServerPlayer.cs:16]: Why?

File: Src/Types/ServerRules.cs [hunk 10]
```diff
@@ -15,17 +15,17 @@ public class ServerRules
         /// <summary>
         /// Mapname
         /// </summary>
-        public string MapName { get; set; }
+        public string MapName { get; set; } = "UNKNOWN";
```
Comment (grish-st) [inline Src/Types/ServerRules.cs:18]: Why?

File: Src/Types/ServerRules.cs [hunk 11]
```diff
@@ -15,17 +15,17 @@ public class ServerRules
...
 
         /// <summary>
         /// Server version
         /// </summary>
-        public string Version { get; set; }
+        public string Version { get; set; } = "UNKNOWN";
```
Comment (grish-st) [inline Src/Types/ServerRules.cs:23]: Why?

File: Src/Types/ServerRules.cs [hunk 12]
```diff
@@ -15,17 +15,17 @@ public class ServerRules
...
 
         /// <summary>
         /// The version of Client Anti-Cheat, for SAMPCAC-enabled servers
         /// </summary>
-        public string SAMPCAC_Version { get; set; } = null;
+        public string SAMPCAC_Version { get; set; } = "N/A";
```
Comment (grish-st) [inline Src/Types/ServerRules.cs:28]: Why?

File: Src/Types/ServerRules.cs [hunk 13]
```diff
@@ -35,7 +35,7 @@ public class ServerRules
         /// <summary>
         /// Link to the server's web page
         /// </summary>
-        public Uri Weburl { get; set; }
+        public Uri Weburl { get; set; } = new Uri("https://sa-mp.com");
```
Comment (grish-st) [inline Src/Types/ServerRules.cs:38]: Why?

File: Src/SampQuery.cs [hunk 14]
```diff
@@ -272,6 +279,54 @@ public async Task<ServerInfo> GetServerInfoAsync()
...
+                return true;
+            }
+            catch
+            {
+                // timeout results in an exception.
+                // a timeout means the server is not open.mp
+                return false;
+            }
+        }
+        /// <summary>
```
Comment (grish-st) [inline Src/SampQuery.cs:329]: I think you can use `SendSocketToServer`

File: Src/SampQuery.cs (hunk 1 above)
Comment (grish-st) [reply Src/SampQuery.cs:455]: Also we need the unit tests for `GetServerIsOMP` as well

**Review by grish-st — CHANGES_REQUESTED** [review]

File: Src/SampQuery.cs (hunk 14 above)
Comment (justmavi) [inline Src/SampQuery.cs:329]: I think you can use SendSocketToServer

File: Src/SampQuery.cs (hunk 1 above)
Comment (justmavi) [inline Src/SampQuery.cs:455]: This block should be refactored, properties need to be checked by name and not by type. What about TryParseByte, it should be tested and unit tests should be written.

Comment (justmavi) [reply Src/SampQuery.cs:455]: Also we need the unit tests for GetServerIsOMP as well

File: Src/Types/ServerInfo.cs (hunk 2 above)
Comment (justmavi) [inline Src/Types/ServerInfo.cs:11]: Why?

File: Src/Types/ServerInfo.cs (hunk 3 above)
Comment (justmavi) [inline Src/Types/ServerInfo.cs:16]: Why?

File: Src/Types/ServerInfo.cs (hunk 5 above)
Comment (justmavi) [inline Src/Types/ServerInfo.cs:21]: Why?

File: Src/Types/ServerInfo.cs (hunk 4 above)
Comment (justmavi) [inline Src/Types/ServerInfo.cs:26]: Why?

File: Src/Types/ServerInfo.cs (hunk 6 above)
Comment (justmavi) [inline Src/Types/ServerInfo.cs:31]: Why?

File: Src/Types/ServerInfo.cs (hunk 7 above)
Comment (justmavi) [inline Src/Types/ServerInfo.cs:36]: Why?

File: Src/Types/ServerInfo.cs (hunk 8 above)
Comment (justmavi) [inline Src/Types/ServerInfo.cs:41]: Why?

File: Src/Types/ServerPlayer.cs (hunk 9 above)
Comment (justmavi) [inline Src/Types/ServerPlayer.cs:16]: Why?

File: Src/Types/ServerRules.cs (hunk 10 above)
Comment (justmavi) [inline Src/Types/ServerRules.cs:18]: Why?

File: Src/Types/ServerRules.cs (hunk 11 above)
Comment (justmavi) [inline Src/Types/ServerRules.cs:23]: Why?

File: Src/Types/ServerRules.cs (hunk 12 above)
Comment (justmavi) [inline Src/Types/ServerRules.cs:28]: Why?

File: Src/Types/ServerRules.cs (hunk 13 above)
Comment (justmavi) [inline Src/Types/ServerRules.cs:38]: Why?

**Review by justmavi — CHANGES_REQUESTED** [review]

**Comment by justmavi** [comment]
Hello, @markski1 !
Thank you for your PR. I wrote some comments and requested some changes on code, please check.
And also let's discuss, which servers are you sending a request to and getting an exception?

**Comment by justmavi** [comment]
Could you provide any references to `open.mp` query mechanism?

**Comment by markski1** [comment]
Hi, I'll be checking the comments in more detail later in the day

To respond to some of the broader questions:

- The season for "UKNOWN" string is that the version of .NET I was writing on does not allow defining those variables without either making them nullable or assigning them any value on declaration.
- I don't have saved any of the servers that caused this failure, but I can try and run a version that logs the caught exceptions and make a list.
- The query mechanism has no documentation, I'll post a reference to the code I took it from when I review the other stuff, I'd have to find it first

**Comment by markski1** [comment]
I've been busy with work but I gave it some attention today.

I installed the correct .NET versions now, and realize that my changes to defaults and nullable stuff were unnecesary, so I've now undone them.

I made some small tweaks (simple 'using' syntax for writer and such) and made a change to use SendSocketToServer which seems to work on a script.

Currently I am struggling to make the VS test tooling work on my computer for some reason. I must sort that out.

As such, I guess this is my TODO:

- Refactor the block within `CollectServerRulesFromByteArray` to grab types by their name.
- Sort out the test stuff in my VS install.
- Create unit test(s) for TryParseByte.
- Create unit test(s) for GetServerIsOMP.
- Possibly update existing tests (they seem to use server IPs that are no longer online).
- Pass the tests.

**Comment by markski1** [comment]
I will not be finishing this.

**Pull request closed without merging**